### DISCUSSION TITLE
Strategies to reduce GitHub Actions consumption

### DISCUSSION THREAD

**Discussion by fdenardo-within3** [renovatebot/renovate#25122] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Hello, first of all, thanks for this excellent app, it's already helping us a lot!

To give a little context to this question (or ask for guidance): We've just started using Renovate on our project's monorepo and it has some old majors and many minor dependencies yet to be upgraded. The other important thing to note is that the project is using GitHub Actions for the CI/CD for which we have minutes assigned to be used on a monthly basis.

My question is specifically related to the last piece of information I gave, the GitHub Actions consumption. I'm looking for strategies (or help if we're doing something wrong) that we can use to ease up the minute consumption because right now we've tried a few different approaches but we're still having some heavy minute consumption which is causing us to pay a lot on CI/CD minutes just for the app (which we don't want to stop using because... it's awesome!).

### What you tried so far.

Using the docs and a few guides out there, we've started using a basic configuration without a schedule to test how much consumption we would have since we had many deps that needed urgent upgrades:

```
{
  $schema: 'https://docs.renovatebot.com/renovate-schema.json',
  extends: [
    'config:js-app',
    'docker:pinDigests',
    ':automergePr', // Raise a PR before auto-merging
    ':labels(dependencies)', // Add "dependencies" labels to any created PR
    ':maintainLockFilesMonthly', // Update lock files on the first day of the month
    ':noUnscheduledUpdates', // Only update branches when scheduled
    ':prConcurrentLimitNone', // Remove the limit for concurrent open PRs
    ':prHourlyLimit1', // Only one PR per hour will be created
  ],
  packageRules: [
    {
      dependencyDashboardApproval: true,
      description: 'Approve major updates',
      labels: ['major'],
      matchUpdateTypes: ['major'],
    },
    {
      dependencyDashboardApproval: true,
      description: 'Approve digest updates',
      labels: ['digest'],
      matchUpdateTypes: ['digest'],
    },
    {
      automerge: true,
      description: 'Auto-merge minor updates',
      labels: ['minor'],
      matchUpdateTypes: ['minor'],
    },
    {
      automerge: true,
      description: 'Auto-merge patch updates',
      labels: ['patch'],
      matchUpdateTypes: ['patch'],
    },
    {
      automerge: true,
      description: 'Auto-merge pin updates',
      labels: ['pin'],
      matchUpdateTypes: ['pin'],
    },
    {
      automerge: true,
      description: 'Auto-merge pin digest updates',
      labels: ['pinDigest'],
      matchUpdateTypes: ['pinDigest'],
    },
  ],
  platformAutomerge: true,
  vulnerabilityAlerts: {
    automerge: true,
    labels: ['security'],
  },
}
```
This approach was useful at the beginning but of course, noisy and expensive because there were a lot of CI/CD running all the time. Since it was just a test to see how much it consumed, we decided to set our ideal schedule which was a narrow one, to ease these costs. We went from a "no schedule" approach to a "really narrowed" one by adding this 👇🏻 

```
    'schedule:automergeEarlyMondays', // Weekly auto-merge schedule on early Monday mornings (before 4 am)
    'schedule:earlyMondays', // Weekly schedule on early Monday mornings (before 4 am)
```

With the schedule, we were looking for Renovate to run its jobs similarly to how it does when it's disabled (at config level), that it just checks the config, and since it's not enabled, it doesn't do any more work and skips everything. This is not the case when we have a schedule, what we've seen in the logs is that each time a job runs it's checking all the created branches and checking one by one if they're on schedule, and only after that is that it skips it and continues. It doesn't run the CI/CD checks so that's good but it is still consuming minutes by doing those checks.

So, is this the correct way of doing this? Are we doing something wrong? Is anyone with the same behavior dealing with it in a different manner?

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: detectSemanticCommits()
DEBUG: semanticCommits: returning "enabled" from cache
DEBUG: processRepo()
DEBUG: Processing 57 branches: renovate/actions-checkout-4.x, renovate/apollo-graphql-packages, renovate/aws-actions-amazon-ecr-login-2.x, renovate/aws-actions-configure-aws-credentials-4.x, renovate/aws-lambda-8.x, renovate/aws-sdk-js-v3-monorepo, renovate/chalk-5.x, renovate/commander-11.x, renovate/csv-stringify-6.x, renovate/datadog-browser-sdk-monorepo, renovate/docker-build-push-action-5.x, renovate/docker-metadata-action-5.x, renovate/eslint-8.x, renovate/faker-js-faker-8.x, renovate/glob-10.x, renovate/graphql-scalars-1.x, renovate/helmet-7.x, renovate/hookform-resolvers-3.x, renovate/i18next-23.x, renovate/json-schema-to-typescript-13.x, renovate/jsonwebtoken-9.x, renovate/jwks-rsa-3.x, renovate/lock-file-maintenance, renovate/major-graphql-tools-monorepo, renovate/major-graphqlcodegenerator-monorepo, renovate/material-ui-monorepo, renovate/mock-jwks-3.x, renovate/mui-x-date-pickers-5.x, renovate/mui-x-date-pickers-6.x, renovate/node-18.x, renovate/node-20.x, renovate/nodemon-3.x, renovate/opensearch-project-opensearch-2.x, renovate/opensearchproject-opensearch-2.x, renovate/opensearchproject-opensearch-dashboards-2.x, renovate/ora-7.x, renovate/postgres-15.x, renovate/postgres-16.x, renovate/prisma-monorepo, renovate/python-3.x, renovate/ramda-0.x, renovate/react-hook-form-7.x, renovate/react-i18next-13.x, renovate/react-monorepo, renovate/supertest-2.x, renovate/testing-library-dom-8.x, renovate/testing-library-dom-9.x, renovate/testing-library-jest-dom-6.x, renovate/testing-library-react-14.x, renovate/ts-loader-9.x, renovate/typescript-5.x, renovate/typescript-eslint-monorepo, renovate/uuid-9.x, renovate/vite-4.x, renovate/vite-plugin-svgr-4.x, renovate/winston-3.x, renovate/yup-1.x
DEBUG: Calculating hourly PRs remaining
DEBUG: currentHourStart=2023-10-09T17:00:00.000+00:00
DEBUG: PR hourly limit remaining: 1
DEBUG: Calculated maximum PRs remaining this run: 1
DEBUG: PullRequests limit = 1
DEBUG: Calculating hourly PRs remaining
DEBUG: currentHourStart=2023-10-09T17:00:00.000+00:00
DEBUG: PR hourly limit remaining: 1
DEBUG: Calculated maximum branches remaining this run: 1
DEBUG: Branches limit = 1
DEBUG: syncBranchState() (branch="renovate/aws-lambda-8.x")
DEBUG: syncBranchState(): update baseBranchSha (branch="renovate/aws-lambda-8.x")
DEBUG: getBranchPr(renovate/aws-lambda-8.x) (branch="renovate/aws-lambda-8.x")
DEBUG: findPr(renovate/aws-lambda-8.x, undefined, open) (branch="renovate/aws-lambda-8.x")
DEBUG: findPr(renovate/aws-lambda-8.x, undefined, closed) (branch="renovate/aws-lambda-8.x")
DEBUG: Found PR #1358 (branch="renovate/aws-lambda-8.x")
DEBUG: branchExists=false (branch="renovate/aws-lambda-8.x")
DEBUG: dependencyDashboardCheck=undefined (branch="renovate/aws-lambda-8.x")
DEBUG: Check for closed PR because recreating closed PRs is disabled. (branch="renovate/aws-lambda-8.x")
DEBUG: findPr(renovate/aws-lambda-8.x, chore(deps): update dependency @types/aws-lambda to v8.10.124, !open) (branch="renovate/aws-lambda-8.x")
DEBUG: prAlreadyExisted=false (branch="renovate/aws-lambda-8.x")
DEBUG: Checking schedule(before 4am on Monday, null) (branch="renovate/aws-lambda-8.x")
DEBUG: Checking 1 schedule(s) (branch="renovate/aws-lambda-8.x")
DEBUG: Checking schedule "before 4am on Monday" (branch="renovate/aws-lambda-8.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_b": [
          14400
        ],
        "d": [
          2
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}

DEBUG: Package not scheduled (branch="renovate/aws-lambda-8.x")
DEBUG: Skipping branch creation as not within schedule (branch="renovate/aws-lambda-8.x")
```

</details>

**Comment by rarkins** [comment]
What you most want to avoid is _unnecessary_ rebases. Renovate will rebase automerging branches if they fall behind the base branch they're to be merged into.

Probably the best long term way to deal with this is the GitHub merge queue. Are you using it?

**Reply by fdenardo-within3** [reply]
Thanks for the quick response!

I agree with the unnecessary rebases, we are trying to avoid those so it doesn't trigger our CI/CD checks, are you saying that with what we did with the schedule is not enough to achieve that? 

What we've seen in the logs when we have both schedule options on our config is what I briefly mentioned before: after onboarding the repo, it starts checking deps that are outdated or have vulnerabilities, then checks each branch's schedule. If not on it, skip it, if it's on it work (meaning rebase, create, merge a branch if necessary). But I don't see it triggering CI/CD checks, is just doing more work than what we think it should globally out of schedule.

And regarding the merge queue, no, we're not using it right now.

**Reply by rarkins** [reply]
You'll trigger a CI/CD check every time there's a git push on a Renovate branch. So for example if Renovate creates a branch/PR, and then it's merged, there's likely two CI/CD events - one when the renovate branch was created and a second on the main branch after merge. If instead, that Renovate branch got rebased 3 times after creation and before merge then you've triggered 5 CI/CD events instead of 2.

The best way to avoid rebasing is to minimize the number of open, automerge-enabled Renovate PRs at a time, because as soon as one PR gets automerged the remainder need rebasing.

**Reply by fdenardo-within3** [reply]
> You'll trigger a CI/CD check every time there's a git push on a Renovate branch. So for example if Renovate creates a branch/PR, and then it's merged, there's likely two CI/CD events - one when the renovate branch was created and a second on the main branch after merge. If instead, that Renovate branch got rebased 3 times after creation and before merge then you've triggered 5 CI/CD events instead of 2.
> 
> The best way to avoid rebasing is to minimize the number of open, automerge-enabled Renovate PRs at a time, because as soon as one PR gets automerged the remainder need rebasing.

I understand what you're saying but I thought that that should not happen unless it's scheduled if you have `:noUnscheduledUpdates` on. Right?

**Reply by rarkins** [reply]
Maybe but automerge won't work if branches aren't up to date

**Reply by fdenardo-within3** [reply]
> Maybe but automerge won't work if branches aren't up to date

OK, so we're on the same path about that. Fortunately, that's not what's happening here because when the schedule's on it works as expected which is awesome. What's happening here is that we're triggering GH Actions while not on schedule (IMO no updates, creates or merges).

Here's a bit of one of the job's logs that maybe would help to clear this out:

```
DEBUG: detectSemanticCommits()
DEBUG: semanticCommits: returning "enabled" from cache
DEBUG: processRepo()
DEBUG: Processing 57 branches: renovate/actions-checkout-4.x, renovate/apollo-graphql-packages, renovate/aws-actions-amazon-ecr-login-2.x, renovate/aws-actions-configure-aws-credentials-4.x, renovate/aws-lambda-8.x, renovate/aws-sdk-js-v3-monorepo, renovate/chalk-5.x, renovate/commander-11.x, renovate/csv-stringify-6.x, renovate/datadog-browser-sdk-monorepo, renovate/docker-build-push-action-5.x, renovate/docker-metadata-action-5.x, renovate/eslint-8.x, renovate/faker-js-faker-8.x, renovate/glob-10.x, renovate/graphql-scalars-1.x, renovate/helmet-7.x, renovate/hookform-resolvers-3.x, renovate/i18next-23.x, renovate/json-schema-to-typescript-13.x, renovate/jsonwebtoken-9.x, renovate/jwks-rsa-3.x, renovate/lock-file-maintenance, renovate/major-graphql-tools-monorepo, renovate/major-graphqlcodegenerator-monorepo, renovate/material-ui-monorepo, renovate/mock-jwks-3.x, renovate/mui-x-date-pickers-5.x, renovate/mui-x-date-pickers-6.x, renovate/node-18.x, renovate/node-20.x, renovate/nodemon-3.x, renovate/opensearch-project-opensearch-2.x, renovate/opensearchproject-opensearch-2.x, renovate/opensearchproject-opensearch-dashboards-2.x, renovate/ora-7.x, renovate/postgres-15.x, renovate/postgres-16.x, renovate/prisma-monorepo, renovate/python-3.x, renovate/ramda-0.x, renovate/react-hook-form-7.x, renovate/react-i18next-13.x, renovate/react-monorepo, renovate/supertest-2.x, renovate/testing-library-dom-8.x, renovate/testing-library-dom-9.x, renovate/testing-library-jest-dom-6.x, renovate/testing-library-react-14.x, renovate/ts-loader-9.x, renovate/typescript-5.x, renovate/typescript-eslint-monorepo, renovate/uuid-9.x, renovate/vite-4.x, renovate/vite-plugin-svgr-4.x, renovate/winston-3.x, renovate/yup-1.x
DEBUG: Calculating hourly PRs remaining
DEBUG: currentHourStart=2023-10-09T17:00:00.000+00:00
DEBUG: PR hourly limit remaining: 1
DEBUG: Calculated maximum PRs remaining this run: 1
DEBUG: PullRequests limit = 1
DEBUG: Calculating hourly PRs remaining
DEBUG: currentHourStart=2023-10-09T17:00:00.000+00:00
DEBUG: PR hourly limit remaining: 1
DEBUG: Calculated maximum branches remaining this run: 1
DEBUG: Branches limit = 1
DEBUG: syncBranchState() (branch="renovate/aws-lambda-8.x")
DEBUG: syncBranchState(): update baseBranchSha (branch="renovate/aws-lambda-8.x")
DEBUG: getBranchPr(renovate/aws-lambda-8.x) (branch="renovate/aws-lambda-8.x")
DEBUG: findPr(renovate/aws-lambda-8.x, undefined, open) (branch="renovate/aws-lambda-8.x")
DEBUG: findPr(renovate/aws-lambda-8.x, undefined, closed) (branch="renovate/aws-lambda-8.x")
DEBUG: Found PR #1358 (branch="renovate/aws-lambda-8.x")
DEBUG: branchExists=false (branch="renovate/aws-lambda-8.x")
DEBUG: dependencyDashboardCheck=undefined (branch="renovate/aws-lambda-8.x")
DEBUG: Check for closed PR because recreating closed PRs is disabled. (branch="renovate/aws-lambda-8.x")
DEBUG: findPr(renovate/aws-lambda-8.x, chore(deps): update dependency @types/aws-lambda to v8.10.124, !open) (branch="renovate/aws-lambda-8.x")
DEBUG: prAlreadyExisted=false (branch="renovate/aws-lambda-8.x")
DEBUG: Checking schedule(before 4am on Monday, null) (branch="renovate/aws-lambda-8.x")
DEBUG: Checking 1 schedule(s) (branch="renovate/aws-lambda-8.x")
DEBUG: Checking schedule "before 4am on Monday" (branch="renovate/aws-lambda-8.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_b": [
          14400
        ],
        "d": [
          2
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}

DEBUG: Package not scheduled (branch="renovate/aws-lambda-8.x")
DEBUG: Skipping branch creation as not within schedule (branch="renovate/aws-lambda-8.x")
```

This section is repeated for each branch. Would any other extract from the job's logs be helpful to give more context to what I'm saying?

**Reply by rarkins** [reply]
I'm not sure what you're asking for, sorry

**Reply by fdenardo-within3** [reply]
> I'm not sure what you're asking for, sorry

Each time Renovate is not on schedule, we can see its jobs run and perform the checks shown in the logs I posted above for each branch it created when it was on schedule. 

Is there a way to skip those checks (when not on schedule) like when Renovate is disabled? Because that's consuming GH Action time which costs money.

**Reply by rarkins** [reply]
Your goal would be to skip the repo processing entirely if outside schedule? e.g. even if branches are conflicted, etc. - don't check and exit immediately?

**Reply by fdenardo-within3** [reply]
> Your goal would be to skip the repo processing entirely if outside schedule? e.g. even if branches are conflicted, etc. - don't check and exit immediately?

Unless I'm getting something wrong, it's already not doing any of that when outside of schedule, It won't rebase it if not on schedule as expected. When on schedule it will, and it will perform all the proper checks as expected so... yeah, shouldn't it only globally check that it's not on schedule and just skip reviewing each branch?

**Reply by rarkins** [reply]
If not on schedule it will still:
- Deconflict conflicted branches,
- Honor any ticked checkboxes in dependency dashboard or PRs (i.e. manual override of schedule)

So if it exits immediately, it can't do either of those, which might annoy users of those repos if they don't understand why.

That said, we certainly could add functionality to skip the repo completely if outside schedule on an opt-in basis

**Reply by fdenardo-within3** [reply]
> If not on schedule it will still:
> 
> * Deconflict conflicted branches,
> * Honor any ticked checkboxes in dependency dashboard or PRs (i.e. manual override of schedule)
> 
> So if it exits immediately, it can't do either of those, which might annoy users of those repos if they don't understand why.
> 
> That said, we certainly could add functionality to skip the repo completely if outside schedule on an opt-in basis

Ah! Gotcha... then that's why all of that branch check is happening, I see, thanks for the explanation. 

Now from what you say, I understand that currently, there's no workaround for that. If that's the case, then the option could be an interesting feature for those like us who want to exit immediately after checking it's not on schedule.

**Reply by viceice** [reply]
you shouldn't worry about the runtime of the hosted app, that won't count to your GitHub actions times. when the app pushes new commits then those can cause action trigger and use your actions times.
